lib/dataset/include/scipp/dataset/sized_dict.h

```diff
@@ -149,6 +173,169 @@ template <class Key, class Value> class SizedDict {
   bool m_readonly{false};
 };
 
+template <class Value> struct AlignedValue {
+  Value value;
+  bool aligned;
+
+  AlignedValue(Value val) noexcept : value(std::move(val)), aligned(true) {}
+  AlignedValue(Value val, const bool a) noexcept
+      : value(std::move(val)), aligned(a) {}
+
+  operator Value() const { return value; }
+
+  bool operator==(const AlignedValue &other) const noexcept {
+    return value == other.value && aligned == other.aligned;
+  }
+
+  bool operator!=(const AlignedValue &other) const noexcept {
+    return !(*this == other);
+  }
+
+  [[nodiscard]] bool is_same(const AlignedValue &other) const noexcept {
+    return value.is_same(other.value) && aligned == other.aligned;
+  }
+
+  [[nodiscard]] bool equals_nan(const AlignedValue &other) const noexcept {
+    return equals_nan(value, other.value) && aligned == other.aligned;
+  }
+
+  [[nodiscard]] auto dim() const -> decltype(auto) { return value.dim(); }
+
+  [[nodiscard]] const core::Dimensions &dims() const noexcept {
+    return value.dims();
+  }
+
+  [[nodiscard]] AlignedValue slice(const Slice &params) const {
+    return {value.slice(params), aligned};
+  }
+
+  void setSlice(const Slice &s, const Variable &data) {
+    value.setSlice(s, data);
+  }
+
+  [[nodiscard]] AlignedValue
+  rename_dims(const std::vector<std::pair<Dim, Dim>> &names,
+              const bool fail_on_unknown = true) {
+    return value.rename_dims(names, fail_on_unknown);
+  }
+
+  [[nodiscard]] AlignedValue as_const() const noexcept {
+    return {value.as_const(), aligned};
+  }
+
+  [[nodiscard]] bool is_readonly() const noexcept {
+    return value.is_readonly();
+  }
+
+  friend std::string to_string(const AlignedValue &x) {
+    return to_string(x.value);
+  }
+};
+
+/// Dict with fixed dimensions and alignment flag.
+template <class Key, class Value>
+class AlignedDict
+    : public SizedDict<Key, AlignedValue<Value>, AlignedDict<Key, Value>> {
```

Comment:
Overall design comment/question. We now have:

- `SizedDict`
- `AlignedDict`, inheriting it, but nevertheless it has to re-implement many methods?
- `AlignedValue`
- Code that needs special handling for aligned coords, even though it does not care about alignment (example: `copy_map` in `operations.cpp`).

This feels wrong, somehow?

- If we have `AlignedValue`, why do we need `AlignedDict`, shouldn't the values handle it? I know, it has to convert to `Variable` in its iteration methods, but....
- Why can't we avoid exposing details such as alignment-flag handling to code that does not have to modify alignment?

What if we think about this differently? Assume we just had `SizedDict` and `AlignedValue`. Which code would not work, if we just use `SizesDict<AlignedValue>` for `Coords`? For the bits that do not, can we make some sort of `Adapter` class (instead of a subclass) that "strips" the aligned flag, i.e., gives you plain variables?

Reply:
I agree. This is a lot of code to do relatively little. The implicit conversion of `AlignedValue` to/from `Variable` should help with many cases. But only if code is explicitly typed to require `Variable`. In generic code, `AlignedValue` can be passed down to some low level code that cannot handle it.

The Python bindings would need to have a special case to expose individual variables. But that should not be too difficult.

I can try this and see what happens.

Not sure how you would implement such an adapter. Do you have something concrete in mind?

Reply:
Let us chat when I am in the office. I don't have a too concrete idea either yet.

Reply:
OK